### PR TITLE
New `BadgeNumber` struct to handle badging - an alternative to using an *int

### PR DESCRIPTION
**Pull request by themartorana**
Related to #38 - you told me to do it so I did ;)

Has new `BadgeNumber` struct and methods, along with tests and an updated `MarshalJSON` method for type `APS`.

See #38 for reasoning.

### CODE REVIEW

**Comment by themartorana** [comment]
Closing and compressing all my bad work into one good commit.

File: badge_number.go [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+)
+
+// BadgeNumber is much a NullInt64 in
+// database/sql except instead of using
+// the nullable value for driver.Value
+// encoding and decoding, this is specifically
+// meant for JSON encoding and decoding
+type BadgeNumber struct {
+	Number int
+	Valid  bool
```
Comment (tylrtrmbl) [inline badge_number.go]: I prefer `isSet`, anyone else?

File: example/example.go [hunk 2]
```diff
@@ -28,7 +28,7 @@ func main() {
 
 		p := apns.NewPayload()
 		p.APS.Alert.Body = body
-		p.APS.Badge = &badge
+		p.APS.Badge = apns.NewBadgeNumber(badge)
```
Comment (tylrtrmbl) [inline example/example.go]: What about this?

```go
p.APS.Badge.Set(badge)
```

I'm trying to get rid of the `NewBadgeNumber` API. Usually `New*` methods return pointers, and I think we can get away entirely without pointers. I think `Set()` will suffice any place we use `NewBadgeNumber`.

File: badge_number.go [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+	b.Number = number
+	b.Valid = true
+	return nil
+}
+
+// MarshalJSON will marshall the numerical value of
+// BadgeNumber
+func (b BadgeNumber) MarshalJSON() ([]byte, error) {
+	return []byte(strconv.Itoa(b.Number)), nil
+}
```
Comment (tylrtrmbl) [inline badge_number.go]: How about `return json.Marshal(b.Number)`?

File: badge_number.go [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+func (b *BadgeNumber) Set(number int) error {
+	if number < 0 {
+		return errors.New("Number must be >= 0")
+	}
+
+	b.Number = number
+	b.Valid = true
+	return nil
+}
+
```
Comment (tylrtrmbl) [inline badge_number.go]: If we make this method take a [`uint`](https://golang.org/ref/spec#Numeric_types) we can get away without the `error` return type (and make this method safe at compile time).

File: badge_number.go [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+		// Since the point of this type is to
+		// allow proper inclusion of 0 for int
+		// types while respecting omitempty,
+		// assume that set==true if there is
+		// a value to unmarshal
+		b.Valid = true
+		return nil
+	}
+}
```
Comment (tylrtrmbl) [inline badge_number.go]: Why don't we use the `json` package's `Unmarshal` method and use type assertion to check that it's an unsigned integer?

**Comment by tylrtrmbl** [comment]
Done with my first pass of review- big thank you @themartorana!

**Comment by themartorana** [comment]
Good notes, all. I'll get them implemented and pushed. Thanks!

**Comment by themartorana** [comment]
This was moved to #60 - didn't realize these comments weren't there. In any case, I'm moving over there, and will implement the same suggestions.

File: badge_number.go (hunk 1 above)
Comment (themartorana) [inline badge_number.go]: Leaving this exported, moving from `Valid` to `IsSet`.

File: example/example.go (hunk 2 above)
Comment (themartorana) [inline example/example.go]: Makes it difficult to write inline initialization: 

```
p := apns.NewPayload{
    Badge: ...
}
```

Comment (tylrtrmbl) [inline example/example.go]: I think that I prefer the explicitness in tests (doesn't change field settings with changed implementations), and we don't expect users to initialize payloads this way.

Comment (tylrtrmbl) [inline example/example.go]: So thank you for changing it in #60 (just saw) :wink:

**Pull request closed without merging**